internal/graph/check_fast_path.go

```diff
@@ -0,0 +1,679 @@
+				return nil, i.err
+			}
+			producer.iter = i.iter
+		}
+	}
+	// TODO: in go1.23 compare performance vs slices.Collect
+	// clean up all empty entries that are both done and drained
+	return slices.DeleteFunc(iterProducers, func(entry *iteratorProducer) bool {
+		return entry.done && entry.iter == nil
+	}), nil
+}
+
+func (c *LocalChecker) fastPathDirect(ctx context.Context,
+	req *ResolveCheckRequest) (chan *iteratorMsg, error) {
+	typesys, _ := typesystem.TypesystemFromContext(ctx)
+	ds, _ := storage.RelationshipTupleReaderFromContext(ctx)
+	tk := req.GetTupleKey()
+	objRel := tuple.ToObjectRelationString(tuple.GetType(tk.GetObject()), tk.GetRelation())
+	i, err := checkutil.IteratorReadStartingFromUser(ctx, typesys, ds, req, objRel, nil)
+	if err != nil {
+		return nil, err
+	}
+	iterChan := make(chan *iteratorMsg, 1)
+	if !concurrency.TrySendThroughChannel(ctx, &iteratorMsg{iter: i}, iterChan) {
+		i.Stop() // will not be received to be cleaned up
+	}
+	close(iterChan)
+	return iterChan, nil
+}
+
+func (c *LocalChecker) fastPathComputed(ctx context.Context, req *ResolveCheckRequest, rewrite *openfgav1.Userset) (chan *iteratorMsg, error) {
+	typesys, _ := typesystem.TypesystemFromContext(ctx)
+	tk := req.GetTupleKey()
+	computedRelation := rewrite.GetComputedUserset().GetRelation()
+	rewrittenTupleKey := tuple.NewTupleKey(
+		tk.GetObject(),
+		computedRelation,
+		tk.GetUser(),
+	)
+
+	childRequest := req.clone()
+	childRequest.TupleKey = rewrittenTupleKey
+	objectType := tuple.GetType(tk.GetObject())
+	rel, err := typesys.GetRelation(objectType, computedRelation)
+	if err != nil {
+		return nil, err
+	}
+
+	return c.fastPathRewrite(ctx, childRequest, rel.GetRewrite())
+}
+
+func fastPathUnion(ctx context.Context, iterProducers []*iteratorProducer, iterChan chan *iteratorMsg) {
+	batch := make([]*openfgav1.TupleKey, 0)
+
+	defer func() {
+		// flush
+		if len(batch) > 0 {
+			concurrency.TrySendThroughChannel(ctx, &iteratorMsg{iter: storage.NewStaticTupleKeyIterator(batch)}, iterChan)
+		}
+		close(iterChan)
+		cleanupIteratorProducers(iterProducers)
+	}()
+
+	/*
+		collect iterators from all channels, until all drained
+		start performing union algorithm across the heads, if an iterator is empty, poll once again the producer
+		ask to see if the channel has a new iterator, otherwise consider it done
+	*/
+
+	for len(iterProducers) > 0 {
+		if ctx.Err() != nil {
+			return
+		}
+		var err error
+		iterProducers, err = pollIteratorProducers(ctx, iterProducers)
+		if err != nil {
+			concurrency.TrySendThroughChannel(ctx, &iteratorMsg{err: err}, iterChan)
+			return
+		}
+		allIters := true
+		minKey := ""
+		indexes := make([]int, 0)
+		for idx, producer := range iterProducers {
+			v, err := producer.iter.Head(ctx)
+			if err != nil {
+				if storage.IterIsDoneOrCancelled(err) {
+					producer.iter.Stop()
+					producer.iter = nil
+					allIters = false
+					// we need to ensure we have all iterators at all times
+					break
```

Comment:
**Suggestion:** Could you just continue the outer loop here instead of tracking the `allIters` var?

Reply:
Potentially? But I'm def not the biggest fan of that pattern. If there is consensus from ya'll I could change it.